functionary/train_vision/train.py

```diff
@@ -0,0 +1,539 @@
+            if label in dic:
+                dic[label]["total"] += 1
+                if label == pred:
+                    dic[label]["acc"] += 1
+
+        metrics = {"accuracy": acc_count / total_num}
+        metrics = {
+            "accuracy_first_token": first_token_correct_count / first_token_total_count,
+            "total_number_first_token": first_token_total_count,
+        }
+
+        for token_id, stat in sorted(
+            first_token_label_dic.items(), key=lambda x: -x[1]["total"]
+        )[:5]:
+            token = tokenizer.decode([token_id])
+            metrics[f"accuracy_first_token_{token}"] = stat["correct"] / stat["total"]
+            metrics[f"accuracy_first_token_{token}_total"] = stat["total"]
+
+        for token_id in dic:
+            token = id2token[token_id]
+            total_num = dic[token_id]["total"]
+            acc = -1
+            if total_num > 0:
+                acc = dic[token_id]["acc"] / total_num
+            metrics[f"accuracy_{token}"] = acc
+            metrics[f"accuracy_total_num_{token}"] = total_num
+
+        return metrics
+
+    def compute_metrics(eval_preds):
+        """Computes next-token accuracy and perplexity metrics for evaluation"""
+        predictions = eval_preds.predictions[0][:, :-1]
+        added_labels = eval_preds.predictions[2]
+        labels = added_labels[:, 1:]
+
+        prediction_list, label_list = (
+            predictions.flatten().tolist(),
+            labels.flatten().tolist(),
+        )
+        # Calculate perplexity
+        loss = eval_preds.predictions[1].tolist()
+        loss = sum(loss) / len(loss)
+        perplexity = math.exp(loss)
+
+        metrics = {"perplexity": perplexity}
+        metrics.update(compute_accuracy_metrics(prediction_list, label_list))
+        return metrics
+
+    def collate_examples(features):
+        result = {}
+        first = features[0]
+        for k, v in first.items():
+            if k in ["input_ids", "attention_mask", "labels"]:
+                if isinstance(v, torch.Tensor):
+                    result[k] = torch.stack([f[k] for f in features])
+                elif isinstance(v, np.ndarray):
+                    result[k] = torch.tensor(np.stack([f[k] for f in features]))
+                else:
+                    result[k] = torch.tensor([f[k] for f in features])
+        # aggregate images & image_sizes
+        images = []
+        for feature in features:
+            images.extend(feature["images"])
+
+        image_sizes = [image.size for image in images]
+        if len(images) > 0:
+            image_tensor = process_images(images, image_processor, model.config)
+        else:
+            image_tensor = None
+        # if type(image_tensor) is not list:
+        #    image_tensor = image_tensor.to(model.dtype)
+
+        result["images"] = image_tensor
+        result["image_sizes"] = image_sizes
+        return result
+
+    class CustomizedTrainer(Trainer):
```

Comment:
Could you add more descriptive name than "Customized". Maybe FunctionAccuracyTrackingTrainer or something like that. And let's add one or two sentence of explanation why we customized the trainer.

Reply:
Sure